tools-public/toolpad/resources/queryHeadlessLibrariesDownloads.ts

```diff
@@ -0,0 +1,81 @@
+// Toolpad queries:
+
+const formatDatePart = (datePart) => {
+  return `${datePart < 10 ? "0" : ""}${datePart}`;
+}
+
+const getDateString = (date: Date) => {
+  const month = date.getMonth() + 1;
+  const day = date.getDate();
+  return `${date.getFullYear()}-${month < 10 ? "0" : ""}${month}-${day < 10 ? "0" : ''}${date.getDate()}`
+}
+
+export const getPackages = (inData) => {
+  const packages: string[] = [];
+  Object.keys(inData ?? {}).forEach(packageName => {
+    packages.push(packageName);
+  });
+  return packages;
+}
+
+const getMonthKey = (date: string) => {
+  return date.slice(0, -2) + "01";
+}
+
+export const prepareData = (inData) => {
+  const date = new Date(2022, 6, 1, 0, 0, 0, 0);
+  const today = new Date();
+  const packages = getPackages(inData);
+
+  const monthsData = {};
+
+  while(date < today) {
+    monthsData[getDateString(date)] = {};
+    packages.forEach(packageName => {
+      monthsData[getDateString(date)][packageName] = 0;
+    })
+    date.setMonth(date.getMonth() + 1);
+  }
+
+  packages.forEach(packageName => {
+    Object.keys(inData[packageName]).map(date => {
+      const monthKey = getMonthKey(date);
+      monthsData[monthKey][packageName] += inData[packageName][date];
+    }) 
+  });
```

Comment:
Oh nice, it looks almost like the logic in https://docs.google.com/spreadsheets/d/1FHxZ456e1t_MNJZg9wrq7pLRFzknZ6jMBhFuW1j1Q8o/edit#gid=0

```jsx
function importdownload(start = '2023-01-01', end = '2023-06-01', packagesRange = [['react-dom', '@mui/base']]) {
  const packages = packagesRange[0].filter((cell) => cell !== '');
  console.log('URL', `https://npm-stat.com/api/download-counts?${addQuery({
    package: packages
  })}&from=${start}&until=${end}`);

  var jsondata = UrlFetchApp.fetch(`https://npm-stat.com/api/download-counts?${addQuery({
    package: packages
  })}&from=${start}&until=${end}`);

  let npmStateResponse = JSON.parse(jsondata.getContentText());

  const months = {};

  Object.keys(npmStateResponse).forEach((package) => {
    const downloads = npmStateResponse[package];

    Object.keys(downloads).forEach((date) => {
      const month = date.substring(0, 7);

      months[month] = months[month] || {};
      months[month][package] = months[month][package] || 0;
      months[month][package] += downloads[date];
    });
  });

  const output = [['month', ...packages]];

  Object.keys(months).sort().forEach((month) => {
    output.push([month, ...packages.map((package) => {
      if (package === 'react-dom') {
        return months[month][package];
      }
      return months[month][package] / months[month]['react-dom'];
    })]);
  });

  console.log('output', output);
  return output;
}
```

Reply:
Yep, looks like the same thing 👍 I was adding to add similar page for all products